File: src/pt/up/fe/comp/dot/ir/DotGraph.java
package pt.up.fe.comp.dot.ir;

import java.util.*;

public class DotGraph {
    public static class Edge {
        public String destination;
        public Map<String, String> attributes = new HashMap<>();
        public boolean directed;

        @Override
        public String toString() {
            return (directed ? "directed " : "") + destination + " " + attributes.toString();
        }

        @Override
        public boolean equals(Object rhs) {
            if (rhs instanceof Edge) {
                Edge other = (Edge) rhs;
                return other.destination.equals(destination) && other.attributes.equals(attributes);
            }
            return false;
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("Name: %s\nStrict: %s\nBidirectional: %s\n", this.name, this.strict, this.bidirectional));
        sb.append(this.attributes.toString()).append("\n");
        for (Map.Entry<String, List<Edge>> node : nodes.entrySet()) {
        	sb.append(String.format("  %s %s -> %s\n",
        			node.getKey(), nodesAttributes.get(node.getKey()).toString(),
        			node.getValue().toString()));
        }

        return sb.toString();
    }

    public void addNode(String node) {
        if (!nodes.containsKey(node))
            nodes.put(node, new ArrayList<Edge>());
        
        if (!nodesAttributes.containsKey(node))
        	nodesAttributes.put(node, new HashMap<String, String>());
    }
    
    public void addNodeAttributes(String node, Map<String, String> attributes) {
    	if (nodesAttributes.containsKey(node))
    		nodesAttributes.get(node).putAll(attributes);
    	else
    		nodesAttributes.put(node, attributes);
    }
    
    public void addAttribute(String attr, String value) {
    	attributes.put(attr, value);
    }

    public Set<String> getNodes() {
        return nodes.keySet();
    }

    public boolean hasNode(String node) {
        return nodes.containsKey(node);
    }
    
    public Map<String, String> getNodeAttributes(String node) {
    	return nodesAttributes.get(node);
    }
    
    public String getAttribute(String attr) {
    	return attributes.get(attr);
    }

    public Edge addNewEdge(String node) {
        List<Edge> nodeEdges = nodes.get(node);
        if (nodeEdges == null)
            throw new NoSuchElementException();

        Edge result = new Edge();
        nodeEdges.add(result);

        return result;
    }

    public List<Edge> getNodeEdges(String node) {
        List<Edge> result = nodes.get(node);
        if (result == null)
            throw new NoSuchElementException();
        return result;
    }


    public boolean strict = false;
    public boolean bidirectional = false;
    public String name;
    public Map<String, List<Edge>> nodes = new LinkedHashMap<>();
    public Map<String, Map<String, String>> nodesAttributes = new HashMap<>();
    public Map<String, String> attributes = new HashMap<>();
}

File: autoanalyze/DotGraph.cs
using System.Collections.Generic;
using System.Text;

namespace autoanalyze
{
    public class DotGraph
    {
        public class Edge
        {
            public Edge()
            {
                Attributes = new Dictionary<string, string>();
            }

            protected bool Equals(Edge other)
            {
                return string.Equals(Destination, other.Destination) && Equals(Attributes, other.Attributes);
            }

            public override int GetHashCode()
            {
                unchecked
                {
                    return ((Destination != null ? Destination.GetHashCode() : 0)*397) ^ (Attributes != null ? Attributes.GetHashCode() : 0);
                }
            }

            public string Destination { get; set; }
            public IDictionary<string, string> Attributes { get; private set; }
            public bool Directed { get; set; }

            public override string ToString()
            {
                var sb = new StringBuilder();
                if (Directed)
                    sb.Append("directed ");
                sb.Append(Destination);
                sb.Append("[");
                foreach (var attribute in Attributes)
                {
                    sb.AppendFormat("{0} - {1}, ", attribute.Key, attribute.Value);
                }
                sb.Append("]");
                return sb.ToString();
            }

            public override bool Equals(object obj)
            {
                if (ReferenceEquals(null, obj)) return false;
                if (ReferenceEquals(this, obj)) return true;
                if (obj.GetType() != GetType()) return false;
                return Equals((Edge) obj);
            }
        }

        public override string ToString()
        {
            var sb = new StringBuilder();
            sb.AppendFormat("Name: {0}\nStrict: {1}\nBidirectional: {2}\n",
                Name, Strict, Bidirectional);

            foreach (var node in Nodes)
            {
                sb.AppendFormat("{0} -> [", node.Key);
                foreach (var edge in node.Value)
                    sb.Append(edge + " ");
                sb.AppendLine("]");
            }

            return sb.ToString();
        }

        public bool Strict { get; set; }
        public bool Bidirectional { get; set; }
        public string Name { get; set; }
        public ISet<string> FinalStats { get; private set; }
        public IDictionary<string, IList<Edge>> Nodes { get; private set; }

        public DotGraph()
        {
            FinalStats = new HashSet<string>();
            Nodes = new Dictionary<string, IList<Edge>>();
        }
    }
}